extensions/amp-video-docking/1.0/controls.js

```diff
@@ -0,0 +1,200 @@
+/**
+ * Copyright 2018 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as Preact from '../../../src/preact';
+import {Timeout} from '../0.1/timeout';
+import {VideoEvents} from '../../../src/video-interface';
+import {listen} from '../../../src/event-helper';
+import {useCallback, useEffect, useMemo, useState} from '../../../src/preact';
+import {useStyles} from './dock.jss';
+import objstr from './obj-str';
+
+/**
+ * A single controls set can be displayed at a time on the controls layer.
+ *
+ * These map to their displayable classname portion in the format
+ * `amp-video-docked-control-set-${NAME}`
+ * e.g. `PLAYBACK: 'playback'` gets `amp-video-docked-control-set-playback`
+ * @enum {string}
+ */
+const ControlSet = {
+  // Playback buttons like play/pause, mute and fullscreen.
+  PLAYBACK: 'playback',
+
+  // Single button to scroll back to inline position of the component. Displayed
+  // during ad playback for CTA interaction.
+  SCROLL_BACK: 'scroll-back',
+};
+
+/** @private @const {!Array<!./breakpoints.SyntheticBreakpointDef>} */
+const BREAKPOINTS = [
+  {
+    className: 'amp-small',
+    minWidth: 0,
+  },
+  {
+    className: 'amp-large',
+    minWidth: 300,
+  },
+];
+
+const TIMEOUT = 1200;
+const TIMEOUT_AFTER_INTERACTION = 800;
+
+/**
+ * @param {*} styles
+ * @return {!PreactDef.Renderable}
+ */
+export function Controls({styles, state, handle}) {
+  const classes = useStyles();
+
+  const [isPlaying, setIsPlaying] = useState(false);
+  const [isShown, setIsShown] = useState(false);
+
+  const hideTimeout = useMemo(
+    () =>
+      new Timeout(window, () => setIsShown(false), {
+        setTimeout: setTimeout.bind(window),
+        clearTimeout: clearTimeout.bind(window),
+      }),
+    []
+  );
+
+  const hideAfterTimeout = useCallback(
+    (after = TIMEOUT) => hideTimeout.trigger(after),
+    [hideTimeout]
+  );
+
+  useEffect(() => {
+    if (!handle) {
+      return;
+    }
+    setIsPlaying(!handle?.paused);
+    const unlisteners = [
+      listen(handle, VideoEvents.PLAYING, () => {
+        setIsPlaying(true);
+      }),
+      listen(handle, VideoEvents.PAUSE, () => {
+        setIsPlaying(false);
+      }),
+    ];
+    return () => {
+      unlisteners.forEach((unlisten) => unlisten());
+    };
+  }, [handle]);
```

Comment:
This is rather awkward and doesn't really work.

The `VideoWrapper` defines its imperative handle, but I haven't found a way to pass down the same ref _inside_ the component. `handle` is currently a ref to the `<video>` element, which is incorrect.